html/browsers/browsing-the-web/history-traversal/persisted-user-state-restoration/scroll-restoration-fragment-scrolling-samedoc.html is failing in WebKit

cdumez opened the pull request:
#### 3b0345e4d4e277783583964c2ff3b9697c5aa1cc
<pre>
html/browsers/browsing-the-web/history-traversal/persisted-user-state-restoration/scroll-restoration-fragment-scrolling-samedoc.html is failing in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=245163">https://bugs.webkit.org/show_bug.cgi?id=245163</a>

Reviewed by Brent Fulgham.

The issue was that the loadType would get ignored in FrameLoader::loadItem()
when the navigation is same-document. As a result, m_loadType wouldn&apos;t get
set and FrameLoader::scrollToFragmentWithParentBoundary() would call
isSameDocumentReload() with an inaccurate loadType.

* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/persisted-user-state-restoration/scroll-restoration-fragment-scrolling-samedoc-expected.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadItem):

Canonical link: <a href="https://commits.webkit.org/254493@main">https://commits.webkit.org/254493@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7fb8b0f1f32c9459da3bf048c808ffd4dd1f74ab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89141 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33702 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19968 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98459 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154771 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93150 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32210 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27766 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81501 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92931 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94788 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25575 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76074 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25510 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80436 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80418 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68484 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29992 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14486 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29718 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15453 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3152 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33164 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38406 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31861 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34546 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -3962,9 +3962,10 @@ void FrameLoader::loadItem(HistoryItem& item, HistoryItem* fromItem, FrameLoadTy
-    if (sameDocumentNavigation)
+    if (sameDocumentNavigation) {
+        m_loadType = loadType;
         loadSameDocumentItem(item);
```
I wonder if `loadSameDocumentItem` should take a '`loadType`' argument, just like `loadDifferentDocumentItem`.

cdumez replied:
But then we have to initialize m_loadType at some point. The question is, in which function is it best to initialize m_loadType. I wasn't sure so I just initialized it there.

In the other code path, they're initializing m_loadType *after* navigation policy but there is no navigation policy here it seems.

brentfulgham replied:
Ah, okay. This is fine as-is.

brentfulgham reviewed (approved):
Looks good, though passing 'loadType' to loadSameDocumentItem would be a bit cleaner.

cdumez committed:
html/browsers/browsing-the-web/history-traversal/persisted-user-state-restoration/scroll-restoration-fragment-scrolling-samedoc.html is failing in WebKit
https://bugs.webkit.org/show_bug.cgi?id=245163

Reviewed by Brent Fulgham.

The issue was that the loadType would get ignored in FrameLoader::loadItem()
when the navigation is same-document. As a result, m_loadType wouldn't get
set and FrameLoader::scrollToFragmentWithParentBoundary() would call
isSameDocumentReload() with an inaccurate loadType.

* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/persisted-user-state-restoration/scroll-restoration-fragment-scrolling-samedoc-expected.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadItem):

Canonical link: https://commits.webkit.org/254493@main
